### PR TITLE
0.5.0

### LINKED ISSUE #119: ValueError: Output array has wrong dimensionality
Hello, I have encountered a problem running vector environments for TurnFaucet-v0 and ms1 tasks,  even with sparse reward.

```bash
python -m mani_skill2.examples.demo_vec_env -e TurnFaucet-v0 --reward-mode sparse
```
```
2023-07-28 15:02:12,107 - mani_skill2 - INFO - Monkey patching gym.spaces.Box.__repr__
opts: []
env_kwargs: {}
[2023-07-28 15:02:13.568] [svulkan2] [error] GLFW error: X11: The DISPLAY environment variable is missing
[2023-07-28 15:02:13.568] [svulkan2] [warning] Continue without GLFW.
2023-07-28 15:02:15,828 - mani_skill2 - INFO - RenderServer is running at: localhost:43115
env VecEnv(<TimeLimit<TurnFaucetEnv<TurnFaucet-v0>>>)
Observation space Dict(agent:Dict(qpos:Box(-inf, inf, (9,), float32), qvel:Box(-inf, inf, (9,), float32), base_pose:Box(-inf, inf, (7,), float32)), extra:Dict(tcp_pose:Box(-inf, inf, (7,), float32), target_angle_diff:Box(-inf, inf, (1,), float32), target_joint_axis:Box(-inf, inf, (3,), float32), target_link_pos:Box(-inf, inf, (3,), float32)), camera_param:Dict(base_camera:Dict(extrinsic_cv:Box(-inf, inf, (4, 4), float32), cam2world_gl:Box(-inf, inf, (4, 4), float32), intrinsic_cv:Box(-inf, inf, (3, 3), float32)), hand_camera:Dict(extrinsic_cv:Box(-inf, inf, (4, 4), float32), cam2world_gl:Box(-inf, inf, (4, 4), float32), intrinsic_cv:Box(-inf, inf, (3, 3), float32))), image:Dict(base_camera:Dict(Color:Box(0.0, 1.0, (128, 128, 4), float32), Position:Box(-inf, inf, (128, 128, 4), float32)), hand_camera:Dict(Color:Box(0.0, 1.0, (128, 128, 4), float32), Position:Box(-inf, inf, (128, 128, 4), float32))))
Action space Box(-1.0, 1.0, (8,), float32)
Episode 0
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/opt/conda/lib/python3.9/site-packages/mani_skill2/examples/demo_vec_env.py", line 111, in <module>
    main()
  File "/opt/conda/lib/python3.9/site-packages/mani_skill2/examples/demo_vec_env.py", line 62, in main
    obs = env.reset()
  File "/opt/conda/lib/python3.9/site-packages/mani_skill2/vector/vec_env.py", line 279, in reset
    return self.reset_wait(indices=indices)
  File "/opt/conda/lib/python3.9/site-packages/mani_skill2/vector/vec_env.py", line 273, in reset_wait
    self._update_np_buffer(results, indices)
  File "/opt/conda/lib/python3.9/site-packages/mani_skill2/vector/vec_env.py", line 226, in _update_np_buffer
    return stack_obs(
  File "/opt/conda/lib/python3.9/site-packages/mani_skill2/vector/vec_env.py", line 425, in stack_obs
    ret[key] = stack_obs(_obs, space[key], buffer=_buffer)
  File "/opt/conda/lib/python3.9/site-packages/mani_skill2/vector/vec_env.py", line 425, in stack_obs
    ret[key] = stack_obs(_obs, space[key], buffer=_buffer)
  File "/opt/conda/lib/python3.9/site-packages/mani_skill2/vector/vec_env.py", line 428, in stack_obs
    return np.stack(obs, out=buffer)
  File "<__array_function__ internals>", line 180, in stack
  File "/opt/conda/lib/python3.9/site-packages/numpy/core/shape_base.py", line 433, in stack
    return _nx.concatenate(expanded_arrays, axis=axis, out=out)
  File "<__array_function__ internals>", line 180, in concatenate
ValueError: Output array has wrong dimensionality
```

The same error replicates in the RL Colab after running this cell (modified only ``env_id`` and ``reward_mode``):
```python
from mani_skill2.vector import VecEnv, make as make_vec_env
num_envs = 2 # recommended value for Google Colab. If you have more cores and a more powerful GPU you can increase this
env: VecEnv = make_vec_env(
    "TurnFaucet-v0",
    num_envs,
    obs_mode=obs_mode,
    reward_mode="sparse",
    control_mode=control_mode,
)
env.seed(0)
obs = env.reset()
print("Base Camera RGB:", obs['image']['base_camera']['rgb'].shape)
print("Base Camera RGB device:", obs['image']['base_camera']['rgb'].device)
env.close()
```
```
2023-07-28 14:46:18,376 - mani_skill2 - INFO - RenderServer is running at: localhost:51711

---------------------------------------------------------------------------

ValueError                                Traceback (most recent call last)

[<ipython-input-18-6567b89c585b>](https://localhost:8080/#) in <cell line: 11>()
      9 )
     10 env.seed(0)
---> 11 obs = env.reset()
     12 print("Base Camera RGB:", obs['image']['base_camera']['rgb'].shape)
     13 print("Base Camera RGB device:", obs['image']['base_camera']['rgb'].device)

8 frames

[/usr/local/lib/python3.10/dist-packages/mani_skill2/vector/vec_env.py](https://localhost:8080/#) in reset(self, indices)
    277     def reset(self, indices=None):
    278         self.reset_async(indices=indices)
--> 279         return self.reset_wait(indices=indices)
    280 
    281     def step_async(self, actions: np.ndarray) -> None:

[/usr/local/lib/python3.10/dist-packages/mani_skill2/vector/vec_env.py](https://localhost:8080/#) in reset_wait(self, indices)
    271         self.waiting = False
    272         vec_obs = self._get_torch_observations()
--> 273         self._update_np_buffer(results, indices)
    274         vec_obs.update(deepcopy(self._obs_np_buffer))
    275         return vec_obs

[/usr/local/lib/python3.10/dist-packages/mani_skill2/vector/vec_env.py](https://localhost:8080/#) in _update_np_buffer(self, obs_list, indices)
    224         for i, obs in zip(indices, obs_list):
    225             self._last_obs_np[i] = obs
--> 226         return stack_obs(
    227             self._last_obs_np, self.non_image_obs_space, self._obs_np_buffer
    228         )

[/usr/local/lib/python3.10/dist-packages/mani_skill2/vector/vec_env.py](https://localhost:8080/#) in stack_obs(obs, space, buffer)
    423             _obs = [o[key] for o in obs]
    424             _buffer = None if buffer is None else buffer[key]
--> 425             ret[key] = stack_obs(_obs, space[key], buffer=_buffer)
    426         return ret
    427     elif isinstance(space, spaces.Box):

[/usr/local/lib/python3.10/dist-packages/mani_skill2/vector/vec_env.py](https://localhost:8080/#) in stack_obs(obs, space, buffer)
    423             _obs = [o[key] for o in obs]
    424             _buffer = None if buffer is None else buffer[key]
--> 425             ret[key] = stack_obs(_obs, space[key], buffer=_buffer)
    426         return ret
    427     elif isinstance(space, spaces.Box):

[/usr/local/lib/python3.10/dist-packages/mani_skill2/vector/vec_env.py](https://localhost:8080/#) in stack_obs(obs, space, buffer)
    426         return ret
    427     elif isinstance(space, spaces.Box):
--> 428         return np.stack(obs, out=buffer)
    429     else:
    430         raise NotImplementedError(type(space))

/usr/local/lib/python3.10/dist-packages/numpy/core/overrides.py in stack(*args, **kwargs)

[/usr/local/lib/python3.10/dist-packages/numpy/core/shape_base.py](https://localhost:8080/#) in stack(arrays, axis, out)
    431     sl = (slice(None),) * axis + (_nx.newaxis,)
    432     expanded_arrays = [arr[sl] for arr in arrays]
--> 433     return _nx.concatenate(expanded_arrays, axis=axis, out=out)
    434 
    435 

/usr/local/lib/python3.10/dist-packages/numpy/core/overrides.py in concatenate(*args, **kwargs)

ValueError: Output array has wrong dimensionality
```

### LINKED ISSUE #124: Change of an asset size after vectorization
I encountered a strange behavior with the `TurnFaucet-v0` environment. When I wrap it in a vector environment, the camera configurations seem to change, resulting in significantly different images from the base and hand cameras (either the camera is too close to the faucet or the faucet appears much larger than the robot).

Here's the code to recreate the issue:

```python
import matplotlib.pyplot as plt
from mani_skill2.vector import make

# Create the vectorized environment
venv = make('TurnFaucet-v0', 2, obs_mode='rgbd', reward_mode='sparse')

# Reset the environment and get the initial observation
obs = venv.reset()

# Close the environment
venv.close()

# Display the image from the base camera
plt.imshow(obs['image']['base_camera']['rgb'].cpu().numpy()[1])
```
image

Upon investigating the `mani_skill2/vector/registration.py` file, I discovered that the `env_fn` function generates the environment with correct images. However, both `RGBDVecEnv` and `VecEnv` produce environments with distorted ones.

Interestingly, the `LiftCube-v0` environment works perfectly fine.

Please note that to run the vectorized `TurnFaucet-v0` environment, you need to use the 0.5.0 branch or visit this GitHub repository: https://github.com/Yefanov/mani-skill2/tree/fix_vec_env

### LINKED ISSUE #125: TurnFaucet-v0 and other envs with mesh sample based rewards are not deterministic.
Currently a Trimesh object is sampled for points to compute rewards for envs like TurnFaucet-v0. The code is currently not doing this by an episode seed but using operating system entropy.

### PR DESCRIPTION
**Pull request by StoneT2000**
Draft PR for upgrading ManiSkill2 to gymnasium as well making various improvements as discussed with @Jiayuan-Gu

Changes:

- [x] upgrade API to gymnasium
- [x] improve reward functions by scaling to [0, 1] and verifying they are working (@xuanlinli17 )
- [ ] Removing goal visuals
- [ ] Updating environment state representations
- [x] update stable baselines notebook tutorials to gymnasium versions
- [x] update stable baselines scripts to gymnasium versions
- [ ] Add CleanRL style baselines
- [x] Semi-automated pytests on all environments (@StoneT2000 )
- [ ] Move downloadable files to Google Storage and avoid using Google Drive.

## Breaking Changes
- `env.render` now accepts no arguments. The old render functions are separated out as other functions and `env.render` calls them and chooses which one based on the `env.render_mode` attribute (set usually upon env creation).
- `env.step` returns `observation, reward, terminated, truncated, info`. See https://gymnasium.farama.org/content/migration-guide/#environment-step for details. For ManiSkill2, the old done signal is now called terminated and truncated is False. All environments by default have a 200 max episode steps so truncated=True after 200 steps.
- `env.reset` returns a tuple `observation, info`. For ManiSkill2, info is always an empty dictionary. Moreover, `env.reset` accepts two new keyword arguments: `seed: int, options: dict | None`. Note that `options` is usually used to configure various random settings/numbers of an environment. Previously ManiSkill2 used to use custom keyword arguments such as `reconfigure`. These keyword arguments are still usable but must be passed through an options dict e.g. `env.reset(options=dict(reconfigure=True))`.
- `env.seed` has now been removed in favor of using `env.reset(seed=val)` per the Gymnasium API.
- ManiSkill VectorEnv is now also modified to adhere to the [Gymnasium Vector Env API](https://gymnasium.farama.org/api/vector/). Note this means that `vec_env.observation_space` and `vec_env.action_space` are batched under the new API, and the individual environment spaces are defined as `vec_env.single_observation_space` and `vec_env.single_action_space`
- All reward functions have been changed to be scaled to the range of [0, 1], generally making any value-learning kind of approach more stable and avoiding gradient explosions. On any environment a reward of 1 indicates success as well and is also indicated by the boolean stored in `info["success"]`. The scaled dense rewards are the new default reward function and is called `normalized_dense`. To use the old <0.5.0 ManiSkill2 dense rewards, set `reward_mode` to `dense`. 

### Code
- Environment code come with separated render functions representing the old render modes. There is now `env.render_human` for creating a interactive GUI and viewer, `env.render_rgb_array` for generating RGB images of the current env from a 3rd person perspective, and `env.render_cameras` which renders all the cameras (including rgb, depth, segmentation if available) and compacts them into one rgb image that is returned. Note that human and rgb_array are used **only** for visualization purposes. They may include artifacts like indicators of where the goal is for visualization purposes, see PickCube-v0 or PandaAvoidObstacles-v0 for examples. cameras mode is reflective of what the actual visual observations are returned by calls to `env.reset` and `env.step`.
- The ManiSkill2 VecEnv creator function `make_vec_env` now accepts a ` max_episode_steps` argument which overrides the default ` max_episode_steps` specified when registering the environment. The default ` max_episode_steps` is 200 for all environments, but note it may be more efficient for RL training and evaluation to use a smaller value as shown in the RL tutorials.

### Tutorials
- All tutorials have been updated to reflect new gym API, new stable baselines 3, and should be more stable on google colab
### Not Code
- New CONTRIBUTING.md document has been added, with details on how to locally develop on ManiSkill2 and test it

## Bug Fixes
- Closes #124 with using the newest version of Sapien, 2.2.2. 
- Closes #119 via #123 where scalar values returned by the state part of a dictionary would cause errors.
- Fixes a compatability bug with Gymnasium AsyncVectorEnv where Gymnasium also could not handle scalar values as it expects shape (1, ), not shape (). This is done by modifying environments to instead of returning floats for certain scalar observation values to return numpy array versions of them. So far only affected TurnFaucet-v0. Partially closes #125 where TurnFaucet-v0 had non-deterministic rewards due to computing rewards based on unseeded sampled points from various meshes. 

## Miscellaneous Changes
- Dockerfile now accepts a python version as an argument
- README and documentation updated to reflect new gym API
- `mani_skill2.examples.demo_vec_env` module now accepts a `--vecenv-type` argument which can be either `ms2` or `gym` and defaults to `ms2`. Lets users benchmark the speed difference themselves. Module was further cleaned to print more nicely
- Various example scripts that have `main` functions now accept an `args` argument and allow for using those scripts from within python and not just the CLI. Used for testing purposes.
- Fix some lack of quietness on some example scripts
- Replaying trajectories accepts a new `--count` argument that lets you specify how many trajectories to replay. There is no data shuffling so the replayed trajectories will always be the same and in the same order. By default this is `None` meaning all trajectories are replayed.

### CODE REVIEW

**Comment by StoneT2000** [comment]
Short update: @tongzhoumu is now contributing to improving some of the reward functions and cleaning up the code so its more usable and effective.

File: requirements.txt [hunk 1]
```diff
@@ -1,9 +1,8 @@
 # core
 numpy<1.24
 scipy
-gym>=0.18.3,<=0.21.0
+gymnasium==0.28.1
```
Comment (Jiayuan-Gu) [inline requirements.txt]: Change to >=0.28.1? I checked the install requirements of `gymnasium`, which looks quite minimal.

File: mani_skill2/__init__.py [hunk 2]
```diff
@@ -56,8 +56,8 @@ def get_commit_info(show_modified_files=False, show_untracked_files=False):
 
 def make_box_space_readable():
     """A workaround for gym>0.18.3,<0.22.0"""
-    import gym
-    from gym import spaces
+    import gymnasium as gym
```
Comment (Jiayuan-Gu) [inline mani_skill2/__init__.py]: This function is not necessary if we upgrade to gymnasium.

File: examples/tutorials/2_reinforcement_learning.ipynb
Comment (StoneT2000) [inline examples/tutorials/2_reinforcement_learning.ipynb:1]: will fix after 0.5.0 is merged

File: examples/tutorials/imitation-learning/bc_liftcube_rgbd.py [hunk 3]
```diff
@@ -50,12 +50,15 @@ def convert_observation(observation):
...
-def rescale_rgbd(rgbd):
+def rescale_rgbd(rgbd, scale_rgb_only=False):
     # rescales rgbd data and changes them to floats
     rgb1 = rgbd[..., 0:3] / 255.0
     rgb2 = rgbd[..., 4:7] / 255.0
-    depth1 = rgbd[..., 3:4] / (2**10)
-    depth2 = rgbd[..., 7:8] / (2**10)
+    depth1 = rgbd[..., 3:4]
+    depth2 = rgbd[..., 7:8]
+    if not scale_rgb_only:
```
Comment (Jiayuan-Gu) [inline examples/tutorials/imitation-learning/bc_liftcube_rgbd.py:59]: You can remove this if-statement if `scale_rgb_only` is always True, or add a comment to explain why depth is normalized by being divided by (2 ** 10).

File: mani_skill2/utils/wrappers/observation.py [hunk 4]
```diff
@@ -2,20 +2,29 @@
...
+from gymnasium import spaces
 
 from mani_skill2.utils.common import (
     flatten_dict_keys,
     flatten_dict_space_keys,
     merge_dicts,
 )
 
+class BaseGymObservationWrapper(gym.ObservationWrapper):
+    """This fixes a particular bug where using the default gym.make_vec_env does not work as it tries to set the spec of an env"""
```
Comment (Jiayuan-Gu) [inline mani_skill2/utils/wrappers/observation.py]: Is it a bug of `gymnasium`? If yes, please include a reported issue or pull request so that we can remove this if the bug is fixed.

File: mani_skill2/utils/registration.py [hunk 5]
```diff
@@ -155,6 +155,7 @@ def _register_env(cls):
             uid,
             entry_point=partial(make, env_id=uid, as_gym=False),
             max_episode_steps=max_episode_steps,
+            disable_env_checker=True,
```
Comment (Jiayuan-Gu) [inline mani_skill2/utils/registration.py]: Is it necessary?

File: mani_skill2/vector/wrappers/sb3.py [hunk 6]
```diff
@@ -23,24 +23,27 @@ def select_index_from_dict(data: dict, i: int):
...
-    """A wrapper for SB3 VecEnv to make it compatible with ManiSkill2 VecEnv."""
+    """A wrapper for ManiSkill2 VecEnv to make it compatible with SB3 VecEnv"""
 
     def __init__(self, venv: VecEnv):
-        super().__init__(venv.num_envs, venv.observation_space, venv.action_space)
+        super().__init__(venv.num_envs, venv.single_observation_space, venv.single_action_space)
         self.venv = venv
+        self._last_seed = None
 
     def seed(self, seed: Optional[int] = None) -> List[Union[None, int]]:
```
Comment (Jiayuan-Gu) [inline mani_skill2/vector/wrappers/sb3.py:35]: Is `seed` still needed?

File: tests/manual_test_venv.py [hunk 7]
```diff
@@ -1,9 +1,11 @@
+# TODO(jigu): This file is incompatible with gymnasium and stable-baselines3
```
Comment (Jiayuan-Gu) [inline tests/manual_test_venv.py]: Is this file still needed given `test_venv.py`?

**Review by Jiayuan-Gu — COMMENTED** [review]
Please also format the directory by black (but please take a look over which files are formatted, since `pyproject.toml` might be updated for filtering which files to be formatted.

File: tests/manual_test_venv.py (hunk 7 above)
Comment (StoneT2000) [reply tests/manual_test_venv.py]: no, removing

File: mani_skill2/vector/wrappers/sb3.py (hunk 6 above)
Comment (StoneT2000) [reply mani_skill2/vector/wrappers/sb3.py:35]: Needed as sb3 uses seed still, despite switching to gymnasium. There is actually small bug in this wrapper that was just fixed but otherwise works as intended.

File: mani_skill2/utils/registration.py (hunk 5 above)
Comment (StoneT2000) [reply mani_skill2/utils/registration.py]: seems not, removing it

Comment (StoneT2000) [reply mani_skill2/utils/registration.py]: actually this creates one bug which is when we use an observation mode of "none", then there is no observation as e.g. MPC does not need.

File: mani_skill2/utils/wrappers/observation.py (hunk 4 above)
Comment (StoneT2000) [reply mani_skill2/utils/wrappers/observation.py]: It is difficult to say if this is a bug of Gymnasium. This is more of our side. We define our own custom registration function that adds extra wrappers. Gymnasium does not expect someone to create their own registration to add custom wrappers, but we do it for conveniency so we can pass in control mode, obs mode in env kwargs. I can change the comment to be more clear.

**Review by Jiayuan-Gu — APPROVED** [review]

**Review by Jiayuan-Gu — APPROVED** [review]
LGTM
